File: 05-Algorithms/code/typescript/src/kmp.ts
/**
 * Builds the KMP "failure function" (a.k.a. prefix function): `pi[i]` is
 * the length of the longest proper prefix of `pat[0..i]` that is also a
 * suffix of it. O(|pat|).
 */
export function kmpPrefixFunction(pat: string): number[] {
  const pi = new Array(pat.length).fill(0);
  let k = 0;
  for (let i = 1; i < pat.length; i++) {
    while (k > 0 && pat[i] !== pat[k]) k = pi[k - 1];
    if (pat[i] === pat[k]) k++;
    pi[i] = k;
  }
  return pi;
}

/**
 * Knuth-Morris-Pratt substring search: returns every start index in `text`
 * where `pat` occurs (overlapping matches included), in ascending order.
 * Returns `[]` if `pat` is empty or longer than `text`. O(|text| + |pat|).
 */
export function kmpSearch(text: string, pat: string): number[] {
  if (pat.length === 0 || pat.length > text.length) return [];
  const pi = kmpPrefixFunction(pat);
  const result: number[] = [];
  let k = 0;
  for (let i = 0; i < text.length; i++) {
    while (k > 0 && text[i] !== pat[k]) k = pi[k - 1];
    if (text[i] === pat[k]) k++;
    if (k === pat.length) {
      result.push(i - k + 1);
      k = pi[k - 1];
    }
  }
  return result;
}

File: 05-Algorithms/code/python/seb_alg/kmp.py
"""Mirrors `typescript/src/kmp.ts` exactly."""

from __future__ import annotations

from typing import List


def kmp_prefix_function(pat: str) -> List[int]:
    """Longest proper prefix-that-is-also-a-suffix length at every index. O(|pat|)."""
    pi = [0] * len(pat)
    k = 0
    for i in range(1, len(pat)):
        while k > 0 and pat[i] != pat[k]:
            k = pi[k - 1]
        if pat[i] == pat[k]:
            k += 1
        pi[i] = k
    return pi


def kmp_search(text: str, pat: str) -> List[int]:
    """KMP substring search: every start index of pat in text, ascending,
    overlaps included. O(|text| + |pat|).
    """
    if len(pat) == 0 or len(pat) > len(text):
        return []
    pi = kmp_prefix_function(pat)
    result: List[int] = []
    k = 0
    for i in range(len(text)):
        while k > 0 and text[i] != pat[k]:
            k = pi[k - 1]
        if text[i] == pat[k]:
            k += 1
        if k == len(pat):
            result.append(i - k + 1)
            k = pi[k - 1]
    return result
